Add UTF-8 design doc stub (#186)
dotnet/designs: accepted/2021/utf8/README.md

# Improve UTF-8 support

**DRAFT**

**PM** [Immo Landwerth](https://github.com/terrajobst)

> UTF-8 is currently the most popular encoding on the web with [more than 95% of all websites using it](https://w3techs.com/technologies/cross/character_encoding/ranking). Additionally it's used as a default encoding on Linux. When .NET was created there was many character encodings used widely and UTF-8 was not dominant at the time - .NET has chose to use UTF-16. As UTF-8 became more popular it become more obvious that UTF-8 has many memory advantages over other encodings including UTF-16 and therefore its great adoption.

Reviewer: In addition to Linux, UTF-8 is also going to become the preferred code page on Windows in a future release.

Reviewer 2: I don't believe that's much of an all-in: the -A APIs internally convert to UTF-16 and then call the appropriate -W API. This is just increasing compatibility with many programs which blindly assumed that -A APIs had UTF-8. I don't expect the native UTF-16 encoding to ever flip.

Reviewer 3: I agree that the Windows OS as a whole is always going to be UTF-16 native. I guess that there may be clusters of new APIs that are UTF-8 native.

Reviewer: > I guess that there may be clusters of new APIs that are UTF-8 native.

This matches my understanding.

Author: I'll leave this as is since I'm not seeing any definite source of truth for what Windows is going to do. The only thing we can say for sure is that they recommend UTF-8 usage

> ## Benefits of using UTF-8 over UTF-16

Reviewer: For completeness, we should also add a drawbacks section. The primary drawbacks are:

* If you perform heavy text processing or utilize globalization-related features heavily (e.g., a GUI application localized to your native language), UTF-16 is often faster than UTF-8 due to technical details of its implementation.
* For string which consist primarily of CJK characters, UTF-8 is less memory-efficient than UTF-16.

These conditions aren't expected to occur frequently in web applications or other application types where TTFB and memory utilization are critical metrics.

> - Reduces heap allocations - in many apps the number of memory needed for storing text is expected to be reduced twice

Reviewer 3: ```suggestion
- Reduces heap allocations - the amount of memory needed for storing text may be reduced up 2x compared to UTF-8 encoding
```

> - Improves quality, readability, and maintainability of code - performance oriented apps frequently store UTF-8 strings as byte arrays or workaround the platform limitations in other ways

Reviewer 2: At the inconvenience of UTF-16 code now dropping in quality, readability and maintainability because of the need to do workarounds being transferred to them.

Author: Adding this to the drawbacks section

> - Improves integration with subsystems that use UTF-8 natively - for example on Linux virtually all APIs take UTF-8 strings - they currently need to transcode from/to UTF-16 - sometimes multiple times - on every API call

Reviewer 3: > on Linux virtually all APIs take UTF-8 strings

Technically, Linux APIs take C strings (ie zero terminated sequences of bytes). UTF-8 encoding for these strings is a convention, but not a requirement. For example, it is easy to create a file with a filename that is invalid UTF-8.

> - Improves integration with subsystems that use UTF-8 natively - for example on Linux virtually all APIs take UTF-8 strings - they currently need to transcode from/to UTF-16 - sometimes multiple times - on every API call

Reviewer 2: WinRT APIs only take UTF-16 strings, so this makes integration with them as well as producing them much harder and need transcoding.

Reviewer 4: There are likewise certain interop structs and APIs on Windows that only take ~~UTF16~~ `WCHAR_T`. The vast majority do have both `A` and `W` overloads, however.

Author: Adding this to the drawbacks section

- Competitivenes with other platforms that offer UTF-8 primitives - most of the popular languages currently offer some sort of UTF-8 string support while .NET only allows transcoding to/from byte array

> - Reduced CPU time for transcoding between UTF-16 and UTF-8 - currently many apps get data as UTF-8 and/or need to produce UTF-8 document output - while transcoding is very fast and in a typical complex app it will not show up very high on the perf metrics it can still add up to couple of percent of CPU cycles

Reviewer 3: It would be useful to have "Data" section with more detailed hard data.

Reviewer 5: @jkotas What do you think about including that in each of the scenarios docs that will be produced so that the performance data section could accompany the specific characteristics of each scenario separately?

Reviewer 3: Sounds good to me.

# Improve UTF-8 support

**DRAFT**

> **PM** [Immo Landwerth](https://github.com/terrajobst)

Reviewer 6: We just removed the PM/Dev markers. Please change this, rebase on top of main, and run `./update-index`  in the root.

```suggestion
**Owner** [Immo Landwerth](https://github.com/terrajobst)
```

> UTF-8 is currently the most popular encoding on the web with [more than 95% of all websites using it](https://w3techs.com/technologies/cross/character_encoding/ranking). Additionally it's used as a default encoding on Linux. When .NET was created there was many character encodings used widely and UTF-8 was not dominant at the time - .NET has chose to use UTF-16. As UTF-8 became more popular it become more obvious that UTF-8 has many memory advantages over other encodings including UTF-16 and therefore its great adoption.

Reviewer 3: ```suggestion
UTF-8 is currently the most popular encoding on the web with [more than 95% of all websites using it](https://w3techs.com/technologies/cross/character_encoding/ranking). Additionally it's used as a default encoding on Linux. When .NET was created there was many character encodings used widely and UTF-8 was not dominant at the time - .NET has chosen to use UTF-16. As UTF-8 became more popular it become more obvious that UTF-8 has many memory advantages over other encodings including UTF-16 and therefore its great adoption.
```

## Benefits of using UTF-8 over UTF-16

- Reduces heap allocations - the amount of memory needed for storing text may be reduced up to 2x compared to UTF-16 encoding
- Improves quality, readability, and maintainability of code - performance oriented apps frequently store UTF-8 strings as byte arrays or workaround the platform limitations in other ways

> - Improves integration with subsystems that are UTF-8 or mostly UTF-8 - for example on Linux most of the APIs are UTF-8 strings are by convention - they currently need to transcode from/to UTF-16 - sometimes multiple times - on every API call

Reviewer 3: ```suggestion
- Improves integration with subsystems that are UTF-8 or mostly UTF-8 - for example on Linux most of the APIs are UTF-8 strings by convention - they currently need to transcode from/to UTF-16 - sometimes multiple times - on every API call
```
